lib/Optimizer/Transforms/MemToReg.cpp

```diff
@@ -820,325 +678,344 @@ class MemToRegPass : public cudaq::opt::impl::MemToRegBase<MemToRegPass> {
+      }
+
+      // We should visit all the blocks at least once.
+      blocksVisited.insert(block);
+      auto preds = dataFlow.getPredecessors(block);
+      for (auto *pred : preds)
+        if (!blocksVisited.count(pred))
+          worklist.push_back(pred);
     } // end of worklist loop
 
-    if (!isFunctionOp(parent)) {
-      // Determine all the unique definitions.
-      SmallVector<Value> allDefs =
-          parent->hasTrait<OpTrait::NoRegionArguments>()
-              ? dataFlow.getLiveOutOfParent()
-              : dataFlow.getAllEscapingBindingDefs();
-
-      if (!allDefs.empty()) {
-        // Replace parent with a copy.
-        SmallVector<Type> resultTypes(parent->getResultTypes());
-        for (auto d : allDefs)
-          resultTypes.push_back(dereferencedType(d.getType()));
-        ConversionPatternRewriter builder(ctx);
-        builder.setInsertionPoint(parent);
-        SmallVector<Value> operands;
-        for (auto opndVal : parent->getOperands())
-          operands.push_back(opndVal);
-        if (!parent->hasTrait<OpTrait::NoRegionArguments>())
-          for (auto d : allDefs)
-            operands.push_back(dataFlow.getPromotedMemRef(d));
-        Operation *np =
-            Operation::create(parent->getLoc(), parent->getName(), resultTypes,
-                              operands, parent->getAttrs(),
-                              parent->getSuccessors(), parent->getNumRegions());
-        builder.insert(np);
-        for (unsigned i = 0; i < parent->getNumRegions(); ++i)
-          builder.inlineRegionBefore(parent->getRegion(i), np->getRegion(i),
-                                     np->getRegion(i).begin());
-        for (unsigned i = 0; i < parent->getNumResults(); ++i)
-          parent->getResult(i).replaceAllUsesWith(np->getResult(i));
-        builder.setInsertionPointAfter(np);
-        for (auto iter : llvm::enumerate(allDefs)) {
-          auto i = iter.index() + parent->getNumResults();
-          if (np->getResult(i).getType() == wireTy)
-            builder.create<quake::WrapOp>(np->getLoc(), np->getResult(i),
-                                          iter.value());
-          else
-            builder.create<cudaq::cc::StoreOp>(np->getLoc(), np->getResult(i),
-                                               iter.value());
-        }
-        cleanUps.insert(parent);
-        parent = np;
+    if (dataFlow.hasLiveOutOfParent()) {
+      // Get all the new results to append.
+      auto allDefs = dataFlow.getLiveOutOfParent();
+
+      // Replace parent with a copy.
+      SmallVector<Type> resultTypes(parent->getResultTypes());
+      for (auto d : allDefs)
+        resultTypes.push_back(dereferencedType(d.getType()));
+      ConversionPatternRewriter builder(ctx);
+      builder.setInsertionPoint(parent);
+      SmallVector<Value> operands;
+      for (auto opndVal : parent->getOperands())
+        operands.push_back(opndVal);
```

Comment:
Maybe I'm missing something, but couldn't this be reduced (like it was done above) to:
```suggestion
      SmallVector<Value> operands(parent->getOperands());
```